File: scripts/migrate_keys_to_key_store.exs
#!/usr/bin/env elixir
# Migration script to import environment variables into KeyManager
#
# Usage:
#   mix run scripts/migrate_keys_to_key_store.exs
#
# This script reads environment variables and stores them in the KeyManager
# for secure key management.

Mix.install([
  {:jason, "~> 1.2"}
])

defmodule KeyMigration do
  @moduledoc """
  Helper module for migrating environment variables to KeyManager
  """

  def run do
    IO.puts("🔐 Key Migration Script")
    IO.puts("=" <> String.duplicate("=", 50))
    IO.puts("")

    # Start the application to initialize KeyManager
    Application.ensure_all_started(:armoricore_realtime)

    migrated = 0
    skipped = 0
    errors = 0

    # JWT Secret
    {migrated, skipped, errors} =
      migrate_key("JWT_SECRET", "jwt.secret", migrated, skipped, errors)

    # FCM API Key
    {migrated, skipped, errors} =
      migrate_key("FCM_API_KEY", "fcm.api_key", migrated, skipped, errors)

    # APNS Keys
    {migrated, skipped, errors} =
      migrate_key("APNS_KEY_ID", "apns.key_id", migrated, skipped, errors)

    {migrated, skipped, errors} =
      migrate_key("APNS_TEAM_ID", "apns.team_id", migrated, skipped, errors)

    {migrated, skipped, errors} =
      migrate_key("APNS_BUNDLE_ID", "apns.bundle_id", migrated, skipped, errors)

    # SMTP Credentials
    {migrated, skipped, errors} =
      migrate_key("SMTP_USERNAME", "smtp.username", migrated, skipped, errors)

    {migrated, skipped, errors} =
      migrate_key("SMTP_PASSWORD", "smtp.password", migrated, skipped, errors)

    # Object Storage Credentials
    {migrated, skipped, errors} =
      migrate_key("OBJECT_STORAGE_ACCESS_KEY", "object_storage.access_key", migrated, skipped, errors)

    {migrated, skipped, errors} =
      migrate_key("OBJECT_STORAGE_SECRET_KEY", "object_storage.secret_key", migrated, skipped, errors)

    IO.puts("")
    IO.puts("=" <> String.duplicate("=", 50))
    IO.puts("Migration Summary:")
    IO.puts("  ✅ Migrated: #{migrated}")
    IO.puts("  ⏭️  Skipped: #{skipped}")
    IO.puts("  ❌ Errors: #{errors}")
    IO.puts("")
    IO.puts("✅ Migration complete!")
    IO.puts("")
    IO.puts("Next steps:")
    IO.puts("  1. Verify keys are stored: mix run -e 'ArmoricoreRealtime.KeyManager.list_keys() |> IO.inspect()'")
    IO.puts("  2. Test key retrieval: mix run -e 'ArmoricoreRealtime.KeyManager.get_jwt_secret(\"jwt.secret\") |> IO.inspect()'")
    IO.puts("  3. Once verified, you can remove environment variables")
  end

  defp migrate_key(env_var, key_id, migrated, skipped, errors) do
    case System.get_env(env_var) do
      nil ->
        IO.puts("⏭️  #{env_var}: Not set, skipping")
        {migrated, skipped + 1, errors}

      value ->
        # Check if key already exists
        if ArmoricoreRealtime.KeyManager.key_exists?(key_id) do
          IO.puts("⏭️  #{env_var} -> #{key_id}: Already exists, skipping")
          {migrated, skipped + 1, errors}
        else
          case ArmoricoreRealtime.KeyManager.store_api_key(key_id, value) do
            :ok ->
              IO.puts("✅ #{env_var} -> #{key_id}: Migrated successfully")
              {migrated + 1, skipped, errors}

            {:error, reason} ->
              IO.puts("❌ #{env_var} -> #{key_id}: Error - #{inspect(reason)}")
              {migrated, skipped, errors + 1}
          end
        end
    end
  end
end

KeyMigration.run()

File: scripts/migrate_keys_to_key_store.rs
//! Migration script to import environment variables into KeyManager (Rust)
//!
//! Usage:
//!   cargo run --bin migrate-keys
//!
//! This script reads environment variables and stores them in the KeyManager
//! for secure key management.

use armoricore_keys::{init_key_store, KeyStore};
use std::env;
use tracing::{info, warn};

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    // Initialize logging
    tracing_subscriber::fmt::init();

    println!("🔐 Key Migration Script");
    println!("{}", "=".repeat(50));
    println!();

    // Initialize key store
    let key_store = init_key_store(None).await?;

    let mut migrated = 0;
    let mut skipped = 0;
    let mut errors = 0;

    // JWT Secret
    let (m, s, e) = migrate_key(&key_store, "JWT_SECRET", "jwt.secret").await;
    migrated += m;
    skipped += s;
    errors += e;

    // FCM API Key
    let (m, s, e) = migrate_key(&key_store, "FCM_API_KEY", "fcm.api_key").await;
    migrated += m;
    skipped += s;
    errors += e;

    // APNS Keys
    let (m, s, e) = migrate_key(&key_store, "APNS_KEY_ID", "apns.key_id").await;
    migrated += m;
    skipped += s;
    errors += e;

    let (m, s, e) = migrate_key(&key_store, "APNS_TEAM_ID", "apns.team_id").await;
    migrated += m;
    skipped += s;
    errors += e;

    let (m, s, e) = migrate_key(&key_store, "APNS_BUNDLE_ID", "apns.bundle_id").await;
    migrated += m;
    skipped += s;
    errors += e;

    // SMTP Credentials
    let (m, s, e) = migrate_key(&key_store, "SMTP_USERNAME", "smtp.username").await;
    migrated += m;
    skipped += s;
    errors += e;

    let (m, s, e) = migrate_key(&key_store, "SMTP_PASSWORD", "smtp.password").await;
    migrated += m;
    skipped += s;
    errors += e;

    // Object Storage Credentials
    let (m, s, e) = migrate_key(
        &key_store,
        "OBJECT_STORAGE_ACCESS_KEY",
        "object_storage.access_key",
    )
    .await;
    migrated += m;
    skipped += s;
    errors += e;

    let (m, s, e) = migrate_key(
        &key_store,
        "OBJECT_STORAGE_SECRET_KEY",
        "object_storage.secret_key",
    )
    .await;
    migrated += m;
    skipped += s;
    errors += e;

    println!();
    println!("{}", "=".repeat(50));
    println!("Migration Summary:");
    println!("  ✅ Migrated: {}", migrated);
    println!("  ⏭️  Skipped: {}", skipped);
    println!("  ❌ Errors: {}", errors);
    println!();
    println!("✅ Migration complete!");
    println!();
    println!("Next steps:");
    println!("  1. Verify keys are stored");
    println!("  2. Test key retrieval");
    println!("  3. Once verified, you can remove environment variables");

    Ok(())
}

async fn migrate_key(
    key_store: &KeyStore,
    env_var: &str,
    key_id: &str,
) -> (usize, usize, usize) {
    match env::var(env_var) {
        Ok(value) => {
            // Check if key already exists
            if key_store.key_exists(key_id).await {
                warn!("{} -> {}: Already exists, skipping", env_var, key_id);
                (0, 1, 0)
            } else {
                match key_store.store_api_key(key_id, &value, None).await {
                    Ok(_) => {
                        info!("{} -> {}: Migrated successfully", env_var, key_id);
                        (1, 0, 0)
                    }
                    Err(e) => {
                        warn!("{} -> {}: Error - {}", env_var, key_id, e);
                        (0, 0, 1)
                    }
                }
            }
        }
        Err(_) => {
            warn!("{}: Not set, skipping", env_var);
            (0, 1, 0)
        }
    }
}

